Fix internal server error when scheduling a wp

b12f opened the pull request:
When rescheduling a work package, in very specific cases the code would
hit the `set_schedule_service`, with a work package without a duration.
The service would attempt to set new wp dates by calculating the end date
from provided start date. However, if the work package has no duration
this calculation fails.

Closes https://community.openproject.org/projects/openproject/work_packages/40921/activity

Commit:
Fix internal server error when scheduling a wp

When rescheduling a work package, in very specific cases the code would
hit the `set_schedule_service`, with a work package without a duration.
The service would attempt to set new wp dates by calculating the end date
from provided start date. However, if the work package has no duration
this calculation fails.

This commit defaults to using a 0 day duration if a duration is not set.

Closes https://community.openproject.org/projects/openproject/work_packages/40921/activity

b12f commented on app/services/work_packages/set_schedule_service.rb:
```diff
@@ -144,9 +144,11 @@ def reschedule_by_predecessors(scheduled, dependency)
     set_dates(scheduled,
               new_start_date,
-              scheduled.due_date && (new_start_date + scheduled.duration - 1))
+              scheduled.due_date && (new_start_date + duration - 1))
```
The alternative would probably be to change this line to the following:

```
scheduled.due_date && !scheduled.duration.nil? && (new_start_date + duration - 1)
```

I'm not sure which is best or more correct.

dombesz replied:
This version would be better, since duration should not ever be 0, it's either nil or bigger than 0.

dombesz replied:
This version would be better, since duration should not ever be 0, it's either nil or bigger than 0.

b12f replied:
Thanks, changed

Commit:
Don't assume a duration of 0 when rescheduling

Instead of rescheduling with a default duration of 0, we check if the
duration is set and if not we do not reschedule the end date.

dombesz reviewed (approved):
LGTM! 👍
